csrc/runtime/fusion_kernel_runtime.cpp

```diff
@@ -361,71 +394,69 @@ void FusionKernelRuntime::compileFusionParallel(KernelArgumentHolder args) {
     }
+  }
 
-    // TODO: index mode should be updated per segmented kernel
-    // Prepare input vector
-    auto group_runtime_inputs =
-        args_manager.translateValsToArgs(group_to_run->inputs());
-    group_runtime_inputs.setDeviceIndex(args.getDeviceIndex());
-    if (group_cache_id.has_value()) {
-      group_runtime_inputs.setCacheId(group_cache_id.value());
-    }
-
-    if (num_groups == 1 || isOptionDisabled(DisableOption::ParallelCompile)) {
-      FUSER_PERF_SCOPE("FusionKernelRuntime::compileFusionParallel");
-      c10::cuda::CUDAGuard dg(args.getDeviceIndex());
-      c10::Device device(c10::DeviceType::CUDA, args.getDeviceIndex());
-      compileKernel(group_runtime_inputs, group_to_run, hic.get());
-    } else {
-      hir::HostIrContainer* hic_ptr = hic.get();
-      // launch compileKernel thread here
-      getThreadPool()->run([this,
-                            args,
-                            group_runtime_inputs,
-                            group_to_run,
-                            &detect_exception_in_thread_pool,
-                            &thread_pool_error_message,
-                            &thread_pool_error_message_mutex,
-                            hic_ptr]() {
-        FUSER_PERF_SCOPE("FusionKernelRuntime::compileFusionParallel");
-        try {
-          c10::cuda::CUDAGuard dg(args.getDeviceIndex());
-          c10::Device device(c10::DeviceType::CUDA, args.getDeviceIndex());
-          compileKernel(group_runtime_inputs, group_to_run, hic_ptr);
-        } catch (const std::exception& e) {
-          // Set flag inside lambda so we can throw an exception after thread
-          // pool completes its work.
-          detect_exception_in_thread_pool.store(true);
-          const std::lock_guard<std::mutex> lock(
-              thread_pool_error_message_mutex);
-          std::stringstream ss;
-          ss << thread_pool_error_message << "\nError from segmentation group "
-             << group_to_run->groupId() << ": " << e.what() << "\n";
-          thread_pool_error_message = ss.str();
-        }
-      });
-    }
+  const std::vector<KernelArgumentHolder> all_runtime_inputs =
+      prepareInputs(args);
 
-    auto fusion_to_run = segmented_fusion_->makeFusion(group_to_run).second;
-    auto group_runtime_outputs =
-        inferOutputSizes(fusion_to_run.get(), group_runtime_inputs);
+  std::atomic<bool> detect_exception_in_thread_pool{false};
+  std::string thread_pool_error_message;
+  std::mutex thread_pool_error_message_mutex;
 
-    // map output args to tensor map
-    args_manager.updateWithSegmentOutputs(
-        group_to_run->outputs(), group_runtime_outputs, run_order_id);
+  // As we pass pointers taken from unique_ptrs to worker threads,
+  // exception handling needs to wait for the worker threads before
+  // cleaning up unique_ptr resources. This try-catch structure is
+  // ugly. Perhaps, we should reconsider passing unique_ptr-backed
+  // data between threads.
+  try {
```

Comment:
After extracting inferOutputSizes to a separate loop, do we still need nested try-catch?

Reply:
Less likely, but still yes since there's a time window where the main thread potentially gets an exception before `wait` is called. I think it's a good practice to remove the potential vulnerable window as much as possible.

More fundamentally, though, I believe it's just a bad idea to pass unique_ptr-backed pointers between threads.